### PR TITLE
fix(tools): split pre-processing, optimization and code validation

### PR DESCRIPTION
**Pull request by NikVolf**
Changes:
- rework of pre-processors
- add `code_validator` module to `gear-wasm-builder`
- split logic into several parts

### CODE REVIEW

**Review by StackOverflowExcept1on — CHANGES_REQUESTED** [review]
we should probably rework the pre-processor and optimizer to fix the original problem

**Comment by breathx** [comment]
> we should probably rework the pre-processor and optimizer to fix the original problem

What is the original problem?

**Comment by NikVolf** [comment]
> > we should probably rework the pre-processor and optimizer to fix the original problem
> 
> What is the original problem?

wasm builder is not very kind to complex custom builds with multiple steps

File: utils/wasm-builder/src/wasm_project.rs [hunk 1]
```diff
@@ -427,26 +430,51 @@ extern "C" fn metahash() {{
...
+                        pre_processed_path,
+                        format!("{file_base_name}_{pre_processor_name}"),
+                    ));
+                }
+            }
+        }
 
-            let mut pre_processed_path = self.wasm_target_dir.clone();
-            pre_processed_path.push(output.path);
+        if wasm_files.is_empty() {
```
Comment (NikVolf) [inline utils/wasm-builder/src/wasm_project.rs:457]: original always should be pushed, why skip it?

or there should be the way for pre-processor to preserve the original compilation artifact under original name

Comment (StackOverflowExcept1on) [reply utils/wasm-builder/src/wasm_project.rs]: done in 454cf15d50072d43d3db7ba8f0395ed5f5a10907

**Review by StackOverflowExcept1on — APPROVED** [review]

**Pull request merged**